packages/components/src/DateTimePickers/pickers/DatePicker/DatePicker.component.js

```diff
@@ -1,83 +1,166 @@
-	const days = (new Array(7))
-					.fill(0)
-					.map((_, i) => i + 1);
+const FIRST_DAY_OF_WEEK = 1;
+const BASE_DATE = new Date(0);
 
-	const weeks = (new Array(4))
+function buildDayNames(firstDayOfweek) {
+	return (new Array(7))
 		.fill(0)
-		.map((_, i) =>
-			days.map(x => ({
-				number: x + (i * 7),
-			}))
-		);
+		.map((_, i) => (i + firstDayOfweek) % 7)
+		.map(dayOfWeek => setDay(BASE_DATE, dayOfWeek))
+		.map(headerDate => format(headerDate, 'dddd'));
+}
+
+const getDayNames = memoize(buildDayNames);
+
+class DatePicker extends React.Component {
+
+	static buildWeeks(year, monthIndex, firstDayOfWeek) {
+		const firstDateOfMonth = new Date(year, monthIndex);
+		const firstDateOfCalendar = startOfWeek(firstDateOfMonth, {
+			weekStartsOn: firstDayOfWeek,
+		});
 
-	const selectedDay = 12;
-	const currentDay = 20;
-	const disabledDays = [6, 15];
+		const lastDateOfMonth = endOfMonth(firstDateOfMonth);
+		const diffWeeks = differenceInCalendarWeeks(lastDateOfMonth, firstDateOfCalendar, {
+			weekStartsOn: firstDayOfWeek,
+		});
+		const nbWeeksToRender = diffWeeks + 1;
 
-	function isSelectedDay(n) {
-		return selectedDay === n;
+		const dates = (new Array(7 * nbWeeksToRender))
+						.fill(0)
+						.map((_, i) => addDays(firstDateOfCalendar, i));
+
+		return chunk(dates, 7);
 	}
 
-	function isCurrentDay(n) {
-		return currentDay === n;
+	constructor(props) {
+		super(props);
+
+		this.getWeeks = memoize(DatePicker.buildWeeks, (year, monthIndex, firstDayOfWeek) => `${year}-${monthIndex}|${firstDayOfWeek}`);
 	}
 
-	function isDisabledDay(n) {
-		return disabledDays.includes(n);
+	isSelectedDate(date) {
+		return this.props.selectedDate !== undefined
+			&& isSameDay(this.props.selectedDate, date);
 	}
 
-	return (
-		<div className={theme.container}>
-			<div className={classNames(theme['calendar-row'], theme['calendar-header-row'])}>
-				{dayNames.map((dayName, i) =>
-					<abbr
-						className={theme['calendar-item']}
-						key={i}
-						title={dayName}
-					>
-						{dayName.charAt(0)}
-					</abbr>
-				)}
-			</div>
+	isDisabledDate(dateToCheck) {
+		const disabledRules = this.props.disabledRules;
+		if (!disabledRules) {
+			return false;
+		}
 
-			<hr className={theme.separator} />
+		return disabledRules
+			.some(disabledRule => {
+				if (typeof disabledRules !== 'object') {
+					return false;
+				}
 
-			{weeks.map((week, i) =>
-				<div
-					className={classNames(theme['calendar-row'], theme['calendar-body-row'])}
-					key={i}
-				>
-					{week.map((day, j) =>
+				switch (disabledRule.constructor) {
+					case Date:
+						return isSameDay(disabledRule, dateToCheck);
+
+					case Array:
+						return disabledRule
+							.some(disabledDateRule => isSameDay(disabledDateRule, dateToCheck));
+
+					default:
+						return false;
+				}
+			});
+	}
+
+	isCurrentMonth(date) {
+		return getMonth(date) === this.props.calendar.monthIndex;
+	}
+
+	render() {
+		const { year, monthIndex } = this.props.calendar;
```

Comment:
I would like to raise an api manipulation issue here.
In `DateTimePicker`, you have `state.calendar`. You destructure it to pass month and year to `DateTimeView`.
In `DateTimeView`, you gather them again to pass a unique prop `calendar`, and pass it to `DatePicker`.
In `DatePicker`, you destructure it again to have month and year.

Can you uniformise please ? Either pass down calendar everywhere, or pass down month and year everywhere.

Reply:
✔

Reply:
you still have a name mismatch
`DateTimePicker`: state.calendar
`DateTimeView`: props.selectedCalendar
`DatePicker`: props.calendar

Reply:
Removed `selected` wording in `DateTimeView`
✔